src/shared/sam/cli/samCliLocalInvoke.ts

```diff
@@ -5,10 +5,113 @@
 
 'use strict'
 
+import * as child_process from 'child_process'
 import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
 import { fileExists } from '../../filesystemUtilities'
-import { DefaultSamCliTaskInvoker } from './samCliInvoker'
-import { SamCliTaskInvoker } from './samCliInvokerUtils'
+import { ChildProcess } from '../../utilities/childProcess'
+import { ChannelLogger } from '../../utilities/vsCodeUtils'
+
+const localize = nls.loadMessageBundle()
+
+export const WAIT_FOR_DEBUGGER_MESSAGES = {
+    PYTHON: 'Waiting for debugger to attach...',
+    NODEJS810: 'Debugger listening on',
+}
+
+export interface SamLocalInvokeCommandArgs {
+    command: string,
+    args: string[],
+    options?: child_process.SpawnOptions,
+    waitForDebuggerAttachMessage: boolean,
+}
+
+/**
+ * Represents and manages the SAM CLI command that is run to locally invoke SAM Applications.
+ */
+export interface SamLocalInvokeCommand {
+    invoke({ }: SamLocalInvokeCommandArgs): Promise<void>
+}
+
+export class DefaultSamLocalInvokeCommand implements SamLocalInvokeCommand {
+    public constructor(private readonly channelLogger: ChannelLogger) {
+    }
+
+    public async invoke({
+        options = {},
+        ...params
+    }: SamLocalInvokeCommandArgs): Promise<void> {
+        this.channelLogger.info(
+            'AWS.running.command',
+            'Running command: {0}',
+            `${params.command} ${params.args.join(' ')}`
+        )
+
+        const childProcess = new ChildProcess(params.command, options, ...params.args)
+
+        await new Promise<void>(async (resolve, reject) => {
+            let checkForDebuggerAttachCue: boolean = params.waitForDebuggerAttachMessage
+
+            // todo : identify the debugger messages to listen for in each runtime
+            const debuggerAttachCues: string[] = [
+                WAIT_FOR_DEBUGGER_MESSAGES.PYTHON,
+                WAIT_FOR_DEBUGGER_MESSAGES.NODEJS810,
+            ]
+
+            await childProcess.start(
```

Comment:
This is the most interesting scenario.

If sam cli does not emit one of the texts we are expecting, code never returns to the caller, the debugger never tries to attach, and things are in a hung state.

I don't know if we can somehow surround this with a (generous) time-out mechanism, but it would make it more robust.

Reply:
We could clean up by killing the `sam local invoke` child process on timeout, which in turn will clean up the docker container. A separate PR might make sense here.

Reply:
Would that leave any hanging threads in the VS Code process?

Reply:
When this PR is merged, I'll write up a separate issue to have a timeout throttle.
I'm adding an onClose check which will address cases where sam local invoke exits when a resolve never occurred. This will clean the code flow in situations where manual intervention took place (eg: manually attaching, or manually killing docker container)